File: src/main/java/com/kgaft/VulkanLib/Device/LogicalDevice/LogicalQueue.java
package com.kgaft.VulkanLib.Device.LogicalDevice;

import com.kgaft.VulkanLib.Utils.DestroyableObject;
import com.kgaft.VulkanLib.Utils.LwjglObject;
import com.kgaft.VulkanLib.Utils.VkErrorException;
import org.lwjgl.PointerBuffer;
import org.lwjgl.vulkan.*;
import static org.lwjgl.vulkan.VK13.*;
import java.lang.instrument.IllegalClassFormatException;

public class LogicalQueue extends DestroyableObject {
    private VkQueue queue;
    private boolean supportPresentation;
    private int queueType;
    private long commandPool;
    private VkDevice device;
    private int index;
    private LwjglObject<VkCommandBufferAllocateInfo> allocInfo;
    private LwjglObject<VkCommandBufferBeginInfo> beginInfo;
    private LwjglObject<VkSubmitInfo> submitInfo;
    public LogicalQueue(VkQueue queue, VkDevice device, boolean supportPresentation, int queueType, int index) throws VkErrorException {
        try
        {
            this.device = device;
            this.queue = queue;
            this.supportPresentation = supportPresentation;
            this.queueType = queueType;
            allocInfo = new LwjglObject<>(VkCommandBufferAllocateInfo.class);
            beginInfo = new LwjglObject<>(VkCommandBufferBeginInfo.class);
            submitInfo = new LwjglObject<>(VkSubmitInfo.class);

            LwjglObject<VkCommandPoolCreateInfo> poolInfo = new LwjglObject<>(VkCommandPoolCreateInfo.class);
            poolInfo.get().queueFamilyIndex(index);
            poolInfo.get().flags(VK_COMMAND_POOL_CREATE_TRANSIENT_BIT|VK_COMMAND_POOL_CREATE_RESET_COMMAND_BUFFER_BIT);
            long[] handleRes = new long[1];
            int status = vkCreateCommandPool(device, poolInfo.get(), null, handleRes);
            if(status!=VK_SUCCESS){
                throw new VkErrorException("Failed to create command pool for queue: "+index+" ", status);
            }
            commandPool = handleRes[0];
        } catch (IllegalClassFormatException e) {
            throw new RuntimeException(e);
        }
    }

    public VkCommandBuffer beginCommandBuffer(){
        allocInfo.get().level(VK_COMMAND_BUFFER_LEVEL_PRIMARY);
        allocInfo.get().commandPool(commandPool);
        allocInfo.get().commandBufferCount(1);
        PointerBuffer pb = PointerBuffer.allocateDirect(1);
        vkAllocateCommandBuffers(device, allocInfo.get(), pb);
        beginInfo.get().flags(VK_COMMAND_BUFFER_USAGE_ONE_TIME_SUBMIT_BIT);
        VkCommandBuffer cmd = new VkCommandBuffer(pb.get(), device);
        pb.free();
        vkBeginCommandBuffer(cmd, beginInfo.get());
        return cmd;
    }


    public void endSingleTimeCommands(VkCommandBuffer commandBuffer) {
        vkEndCommandBuffer(commandBuffer);
        PointerBuffer pb = PointerBuffer.allocateDirect(1);
        pb.put(commandBuffer);
        submitInfo.get().pCommandBuffers(pb);
        vkQueueSubmit(queue, submitInfo.get(), 0);
        vkQueueWaitIdle(queue);
        vkFreeCommandBuffers(device, commandPool, commandBuffer);
        pb.free();
    }

    public VkQueue getQueue() {
        return queue;
    }

    public boolean isSupportPresentation() {
        return supportPresentation;
    }

    public int getQueueType() {
        return queueType;
    }

    public VkDevice getDevice() {
        return device;
    }

    public int getIndex() {
        return index;
    }

    @Override
    public void destroy() {
        destroyed = true;
        vkDestroyCommandPool(device, commandPool, null);
    }
}

File: src/cxx/VulkanLib/Device/LogicalDevice/LogicalQueue.cpp
//
// Created by kgaft on 3/11/24.
//

#include "LogicalQueue.hpp"
#include "VulkanLib/VulkanLibAllocationCallback.h"

LogicalQueue::LogicalQueue(vk::Queue queue, vk::Device device, bool supportPresentation, vk::QueueFlags queueType,
                           unsigned int index) : queue(
        queue),
                                                 supportPresentation(
                                                         supportPresentation),
                                                 queueType(
                                                         queueType),
                                                 device(device), index(index) {
    vk::CommandPoolCreateInfo poolInfo = {};
    poolInfo.queueFamilyIndex = index;
    poolInfo.flags =
            vk::CommandPoolCreateFlagBits::eTransient | vk::CommandPoolCreateFlagBits::eResetCommandBuffer;
    commandPool = device.createCommandPool(poolInfo, VkLibAlloc::acquireAllocCb().get());
}

vk::CommandBuffer LogicalQueue::beginSingleTimeCommands() {
    allocInfo.level = vk::CommandBufferLevel::ePrimary;
    allocInfo.commandPool = commandPool;
    allocInfo.commandBufferCount = 1;

    vk::CommandBuffer commandBuffer = device.allocateCommandBuffers(allocInfo)[0];



    beginInfo.flags = vk::CommandBufferUsageFlagBits::eOneTimeSubmit;

    commandBuffer.begin(beginInfo);
    return commandBuffer;
}

void LogicalQueue::endSingleTimeCommands(vk::CommandBuffer commandBuffer) {
    commandBuffer.end();
    submitInfo.commandBufferCount = 1;
    submitInfo.pCommandBuffers = &commandBuffer;
    queue.submit(submitInfo);
    queue.waitIdle();
    device.freeCommandBuffers(commandPool, commandBuffer);
}

const vk::Queue &LogicalQueue::getQueue() const {
    return queue;
}

bool LogicalQueue::isSupportPresentation() const {
    return supportPresentation;
}

vk::QueueFlags LogicalQueue::getQueueType() const {
    return queueType;
}

const vk::CommandPool &LogicalQueue::getCommandPool() const {
    return commandPool;
}

unsigned int LogicalQueue::getIndex() const {
    return index;
}

void LogicalQueue::destroy() {
    destroyed = true;
    device.destroyCommandPool(commandPool);
}
